numpy/_core/src/multiarray/ctors.c

```diff
@@ -2440,36 +2447,65 @@ PyArray_FromArrayAttr_int(
         return Py_NotImplemented;
     }
 
-    PyObject *copy = never_copy ? Py_False : Py_None;
     PyObject *kwargs = PyDict_New();
-    PyDict_SetItemString(kwargs, "copy", copy);
+
+    /*
+     * Only if the value of `copy` isn't the default one, we try to pass it
+     * along; for backwards compatibility we then retry if it fails because the
+     * signature of the __array__ method being called does not have `copy`.
+     */
+    int copy_passed = 0;
+    if (copy != -1) {
+        copy_passed = 1;
+        PyObject *copy_obj = copy == 1 ? Py_True : Py_False;
+        PyDict_SetItemString(kwargs, "copy", copy_obj);
+    }
     PyObject *args = descr != NULL ? PyTuple_Pack(1, descr) : PyTuple_New(0);
 
     new = PyObject_Call(array_meth, args, kwargs);
 
-    if (PyErr_Occurred()) {
+    if (new == NULL) {
+        if (npy_ma_str_array_err_msg_substr == NULL) {
+            return NULL;
+        }
         PyObject *type, *value, *traceback;
         PyErr_Fetch(&type, &value, &traceback);
-        if (PyUnicode_Check(value) && PyUnicode_CompareWithASCIIString(value,
-                    "__array__() got an unexpected keyword argument 'copy'") == 0) {
-            Py_DECREF(type);
-            Py_XDECREF(value);
-            Py_XDECREF(traceback);
-            if (PyErr_WarnEx(PyExc_UserWarning,
-                             "__array__ should implement 'dtype' and 'copy' keywords", 1) < 0) {
-                return NULL;
+        if (value != NULL) {
+            PyObject *str_value = PyObject_Str(value);
+            if (PyUnicode_Contains(
+                        str_value, npy_ma_str_array_err_msg_substr) > 0) {
+                Py_DECREF(type);
+                Py_DECREF(value);
+                Py_XDECREF(traceback);
+                if (PyErr_WarnEx(PyExc_UserWarning,
+                                 "__array__ should implement 'dtype' and "
+                                 "'copy' keywords", 1) < 0) {
+                    Py_DECREF(str_value);
+                    Py_DECREF(kwargs);
+                    return NULL;
+                }
+                if (copy_passed) { /* try again */
+                    PyDict_DelItemString(kwargs, "copy");
+                    new = PyObject_Call(array_meth, args, kwargs);
+                    if (new == NULL) {
+                        Py_DECREF(str_value);
+                        Py_DECREF(kwargs);
+                        return NULL;
+                    }
+                }
             }
-            Py_SETREF(new, PyObject_Call(array_meth, args, NULL));
-        } else {
+            Py_DECREF(str_value);
+        }
+        if (new == NULL) {
             PyErr_Restore(type, value, traceback);
+            Py_DECREF(kwargs);
             return NULL;
         }
     }
 
+    Py_DECREF(kwargs);
```

Comment:
Should we have `Py_DECREF(args);` next to `kwargs`? Similarly to: https://github.com/numpy/numpy/blob/9d69a62cdc734e9fae074a9242aa1b00eceafcd4/numpy/_core/src/multiarray/arrayfunction_override.c#L240

Reply:
And this is the only comment that I have.